pmd-java/src/main/java/net/sourceforge/pmd/lang/java/symboltable/SourceFileScope.java

```diff
@@ -136,7 +137,12 @@ public ClassNameDeclaration findClassNameDeclaration(String name) {
      * @return set of all types in this source file.
      */
     public Map<String, Node> getQualifiedTypeNames() {
-        return getSubTypes(null, this);
+        if (qualifiedTypeNames != null) {
+            return qualifiedTypeNames;
+        }
+
+        qualifiedTypeNames = getSubTypes(null, this);
```

Comment:
Nevermind, looks like even `AbstractScope` shares its internal state, and makes allocations galore.